### PR TITLE
Add per-installation configuration folder

### PR DESCRIPTION
**Pull request by sam-harry**
First, thank you for your work for the freenas & homeassistant communities!

Please consider adding a per-installation configuration folder with the following changes:

- update motd to point to per-installation configuration file
- add per-installation file with fields similar to homeassistant os
addon
- add per-installation pwfile
- modify plugin configuration file to include per-installation
configuration files

On installation of the mosquitto plugin, I was not sure how to mount the configuration files outside of the jail. If you mount the entire /usr/local/etc/mosquitto directory then the mosquitto binaries are kept, which may cause confusion when updating. Having a separate directory for per-installation configuration files provides a location for the plugin user to mount configuration files into the jail, so that the jail may be destroyed and rebuilt and still preserve their configuration. Users who prefer to edit the mosquitto plugin configuration directly should not be effected.

### CODE REVIEW

**Comment by tprelog** [comment]
@sam-harry - Thank you for your suggestion and pull request. I very much agree with the idea for your purposed changes, however I don't it will work exactly the way you have in mind.

>On installation of the mosquitto plugin, I was not sure how to mount the configuration files outside of the jail.

I think you are looking at this backwards. Think of it like this, we would like to have configuration files saved outside the jail. When we reinstall install the plugin, we actually want to mount (external) configuration files *inside* the jail. In order to mount an external dataset, the mount directory _inside_ the jail needs to _empty_. If we have already provided files inside `/usr/local/etc/mosquitto/conf/` you will not be able to mount your external data-set to this location.

I have tried to stay neutral here (I guess not everybody uses Home Assistant) but I only made these plugins because I use them with HA. That said, I am very much on-board with trying to provide a config setup that reflects the Mosquitto add-on for Home Assistant OS, with just a slight change. Do not include the `conf` directory by default. Instead move `overlay/usr/local/etc/mosquitto/conf/mosquitto.conf` to `overlay/usr/local/etc/mosquitto/mosquitto.conf.homeassistant` - I don't think we need the empty `overlay/usr/local/etc/mosquitto/conf/pwfile` as this file would be generated when adding the MQTT user.

For a Home Assistant OS type config, I might even suggest using `/config/mosquitto` rather then `/usr/local/etc/mosquitto/conf/` but either way the idea is the same.

Please consider this example of how I think of your idea could work:

Assuming you have already installed Mosquitto as normal and it includes your `mosquitto.conf.homeassistant` file

**First time setup to use external, HA OS type configuration**
1. On FreeNAS / TrueNAS create the `config` dataset. This would be the parent dataset to hold all your configuration files for HA and related plugins, in this case mosquitto. This would also be the location you would setup file sharing (I assume samba) service to use.

2. For Mosquitto, we create a child dataset named `mosquitto` inside the `config` dataset previously created.  ( At this point obviously, it's empty )

3. Next using the Jail's -> MOUNT POINTS,  mount the new dataset inside the plugin, to location you have chosen in your `mosquitto.conf.homeassistant` file. There is no need to create this directory _inside_ the jail ahead of time, but if it does exist, it must be empty.

4 . After the dataset is mounted we enter the jails console and copy the `mosquitto.conf.homeassistant` file to the location we have mount the dataset. For example:
```
cp /usr/local/etc/mosquitto/mosquitto.conf.homeassistant /config/mosquitto/mosquitto.conf
```
( probably suggest to configure the file at this point as well)

5. Finally, set the new location for the configuration file and restart mosquito to apply changes
```
sysrc mosquitto_config=/config/mosquitto/mosquitto.conf
service mosquitto restart
```

---

**Reinstalling the Mosquitto plugin after the initial set-up**

Basically repeat only steps 3 and 5 from above, but this time, you can save a step ( from 5 ) by doing them backwards

1. Set the new location for the configuration file you have chosen
```
sysrc mosquitto_config=/config/mosquitto/mosquitto.conf
```

2. Using the Jail's -> MOUNT POINTS, mount your existing `config/mosquitto`  dataset inside the plugin, to the location you have chose.

When you start the jail backup - Mosquitto should be running with your previous configuration

**Comment by sam-harry** [comment]
@tprelog thank you for the well written response. Your how-to may make an excellent wiki addition.

It looks like there are two ways to approach this: using the `include_dir /path/to/config` option in the default config `/usr/local/etc/mosquitto/mosquitto.conf` and by changing the location of the primary configuration file with `sysrc mosquitto_config=/path/to/mosquitto.conf`. In any case, I now see your point against adding an additional conf directory.

My initial goal with changing mosquitto.conf.plugin was to make commonly changed options more readily available to the plugin user, using the homassistant os plugin as a reference for commonly changed options. I do agree that it is better to remain neutral.

With these things in mind I looked at other freenas plugins and found that the use of settings.json may be very useful in exposing commonly changed options. These are not commonly implemented, it seems, however freenas used to expose these options in the legacy UI. Nowadays, they are command line only with an open ticket to re-implement UI options in truenas core 12, but still useful.

So, please consider the most recent changes. Starting with the low hanging fruit, I added iocage set commands that can be used like the following examples:

- `iocage set -P port=1883 mosquitto`
- `iocage set -P allow_anonymous=false mosquitto`
- `iocage set -P adduser=username,password mosquitto`
- `iocage set -P deluser=username mosquitto`

that will change the port setting, allow_anonymous setting, add users to the default password file, and delete users from the default password file, respectively.
As well as iocage get commands that can be used like:

- `iocage get -P port mosquitto`
- `iocage get -P allow_anonymous mosquitto`

to get the currently configured port setting and allow_anonymous setting.

I think that these changes may be useful to any mosquitto plugin user.

Now, the installation process for users how only wish to specify these values (perhaps a small number but they suit my needs, at least, and more options can be added) would look more like:

- install the plugin

then, from the freenas shell (NOT the jail shell)

- `iocage set -P port=1883 mosquitto`
- `iocage set -P adduser=username,password mosquitto`
- `iocage set -P allow_anonymous=false mosquitto`

Plugin user who follow your how-to can change the config directory altered by the iocage set command by specifying _CONFIG and _PWFILE variables in /usr/local/bin/mosquittoset and /usr/local/bin/mosquittoset.

PS I appologize for the disorganized commits, it really should have been just one.

**Comment by tprelog** [comment]
@sam-harry - **Thank you!** This is a huge step forward, one that's been on my "To Do" list for quite some time.

This looks great, I see no reason I can't just merge this now. However, I should actually test the Plugin -> UPGRADE function, just to be safe.

I'm busy in real life for the next few days, so I won't have time until end of the week. I hope to have your changes merged next weekend but I just want to let you know now, I really appreciate this. Thanks again! :smiley: 

---

>PS I appologize for the disorganized commits, it really should have been just one.

No worries my friend. Far as I'm concerned, it's perfectly fine
